alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -0,0 +1,155 @@
+    (P.LowerBound P.NegInf True)
+    (P.UpperBound P.PosInf False)
+transVI (ValidityInterval (SJust (SlotNo i)) SNothing) =
+  P.Interval
+    (P.LowerBound (P.Finite (fromIntegral i)) True)
+    (P.UpperBound P.PosInf False)
+transVI (ValidityInterval SNothing (SJust (SlotNo i))) =
+  P.Interval
+    (P.LowerBound P.NegInf True)
+    (P.UpperBound (P.Finite (fromIntegral i)) False)
+transVI (ValidityInterval (SJust (SlotNo i)) (SJust (SlotNo j))) =
+  P.Interval
+    (P.LowerBound (P.Finite (fromIntegral i)) True)
+    (P.UpperBound (P.Finite (fromIntegral j)) False)
+
+transTxIn' :: CC.Crypto c => TxIn c -> P.TxOutRef
+transTxIn' (TxIn txid nat) = P.TxOutRef (transTxId txid) (fromIntegral nat)
+
+transTxIn :: forall c. CC.Crypto c => TxIn c -> P.TxInInfo
+transTxIn txin =
+  P.TxInInfo
+    (transTxIn' txin)
+    Nothing -- NOT SURE ABOUT THIS
+    (transValue (inject @(Mary.Value c) (Coin 0))) -- OR ABOUT THIS
+
+transTxOut :: forall era. (Era era, Value era ~ Mary.Value (Crypto era)) => TxOut era -> P.TxOut
+transTxOut (TxOut (Addr _network (ScriptHashObj sh) _stake) val datahash) =
+  P.TxOut
+    (P.ScriptAddress (transScriptHash sh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+transTxOut (TxOut (Addr _network (KeyHashObj kh) _stake) val datahash) =
+  P.TxOut
+    (P.PubKeyAddress (transKeyHash kh))
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+transTxOut (TxOut (AddrBootstrap (BootstrapAddress _ad)) val datahash) =
+  P.TxOut
+    (P.PubKeyAddress (P.PubKeyHash undefined)) -- TODO BootstrapAddress into ByteString
+    (transValue @(Crypto era) val)
+    (P.PayToScript (transDataHash datahash))
+
+transValue :: forall c. Mary.Value c -> P.Value
+transValue (Mary.Value _coin _mp) = undefined -- TODO complete this
+
+transKeyHash :: KeyHash d c -> P.PubKeyHash
+transKeyHash (KeyHash (UnsafeHash h)) = P.PubKeyHash (fromShort h)
+
+transScriptHash :: ScriptHash i -> P.ValidatorHash
+transScriptHash (ScriptHash (UnsafeHash h)) = P.ValidatorHash (fromShort h)
+
+transDataHash :: StrictMaybe (DataHash c) -> P.DatumHash
+transDataHash (SJust safe) = P.DatumHash (transSafeHash safe)
+transDataHash SNothing = P.DatumHash "\00" -- TODO THIS IS PROBABLY WRONG
+
+transTx ::
+  forall era.
+  ( Era era,
+    Core.TxBody era ~ Alonzo.TxBody era,
+    Value era ~ Mary.Value (Crypto era)
+  ) =>
+  Tx era ->
+  P.TxInfo
+transTx tx =
+  P.TxInfo
+    (map transTxIn (Set.toList allinputs))
+    (map transTxOut (foldr (:) [] outs))
+    (transValue (inject @(Mary.Value (Crypto era)) fee))
+    (transValue forge)
+    (transVI interval)
+    [] -- MonetaryPolicyHash ???
+    [] -- PubKeyHash ???
+    [] -- Auxiliary Data and hashes
+    (P.TxId (transSafeHash (hashAnnotated @(Crypto era) tx)))
+  where
+    tbody = body' tx
+    _witnesses = wits' tx
+    _isval = isValidating' tx
+    _auxdat = auxiliaryData' tx
+    allinputs = Set.union (inputs' tbody) (inputs_fee' tbody)
```

Comment:
I think we decided we'd list them separately?

Reply:
Split  into two fields.